tools/mtouch/Application.cs

```diff
@@ -1048,131 +1423,146 @@ public static void CopyMSymData (string src, string dest)
-					all_frameworks.Add (Path.Combine (Driver.GetProductSdkDirectory (this), "Frameworks", "Mono.framework"));
-				}
-				
-				foreach (var appex in Extensions) {
-					var f_path = Path.Combine (appex, "..", "frameworks.txt");
-					if (!File.Exists (f_path))
+			}
+
+			// And from ourselves
+			var all_assemblies = Targets.SelectMany ((v) => v.Assemblies);
+			var all_frameworks = Frameworks.Concat (all_assemblies.SelectMany ((v) => v.Frameworks));
+			var all_weak_frameworks = WeakFrameworks.Concat (all_assemblies.SelectMany ((v) => v.WeakFrameworks));
+			foreach (var fw in all_frameworks.Concat (all_weak_frameworks)) {
+				BundleFileInfo info;
+				if (!Path.GetFileName (fw).EndsWith (".framework", StringComparison.Ordinal))
+					continue;
+				var key = $"Frameworks/{Path.GetFileName (fw)}";
+				if (!bundle_files.TryGetValue (key, out info))
+					bundle_files [key] = info = new BundleFileInfo ();
+				info.Sources.Add (fw);
+			}
+
+			// We also need to add any frameworks from extensions
+			foreach (var appex in AppExtensions) {
+				foreach (var bf in appex.bundle_files.ToList ()) {
+					if (!Path.GetFileName (bf.Key).EndsWith (".framework", StringComparison.Ordinal) && !bf.Value.DylibToFramework)
 						continue;
 
-					foreach (var fw in File.ReadAllLines (f_path)) {
-						Driver.Log (3, "Copying {0} to the app's Frameworks directory because it's used by the extension {1}", fw, Path.GetFileName (appex));
-						all_frameworks.Add (fw);
-					}
+					Driver.Log (3, "Copying {0} to the app's Frameworks directory because it's used by the extension {1}", bf.Key, Path.GetFileName (appex.Name));
+
+					var appex_info = bf.Value;
+					BundleFileInfo info;
+					if (!bundle_files.TryGetValue (bf.Key, out info))
+						bundle_files [bf.Key] = info = new BundleFileInfo ();
+					info.Sources.UnionWith (appex_info.Sources);
+					if (appex_info.DylibToFramework)
+						info.DylibToFramework = true;
 				}
+			}
 
-				foreach (var fw in all_frameworks) {
-					if (!fw.EndsWith (".framework", StringComparison.Ordinal))
-						continue;
-					if (!Xamarin.MachO.IsDynamicFramework (Path.Combine (fw, Path.GetFileNameWithoutExtension (fw)))) {
-						// We can have static libraries camouflaged as frameworks. We don't want those copied to the app.
-						Driver.Log (1, "The framework {0} is a framework of static libraries, and will not be copied into the app.", fw);
-						continue;
+			// Finally copy all the files & directories
+			foreach (var kvp in bundle_files) {
+				var name = kvp.Key;
+				var info = kvp.Value;
+				var targetPath = Path.Combine (AppDirectory, name);
+				var files = info.Sources;
+				var isFramework = Directory.Exists (files.First ());
+
+				if (IsExtension && !IsWatchExtension && (isFramework || info.DylibToFramework))
+					continue; // Don't copy frameworks to app extensions (except watch extensions), they go into the container app.
```

Comment:
You mean intents extensions and watchkit extensions reference frameworks that we put into the watch app?

I think we can implement support for intents extensions easier than that:

* We keep frameworks in the watchkit extension.
* Intents extensions reference the frameworks from their container (the fact that it's a container .appex instead of container .app doesn't require any code changes, currently it's referenced as `@executable_path/../../Frameworks`, which works when the container is either an .appex or .app).
* We build watchkit extensions like an .app with extensions (there are very few, if any, broken assumptions about this in mtouch right now, and if there are, they should be trivial to fix).

For terminology I suggest:

* **WatchKit extension** for what we're currently calling *Watch extension* in some places.
* Never use **Watch extension** anywhere.
* **watchOS extension** for any extension that can be found inside a watchOS app. Currently that would include both **WatchKit extensions** and **Intents extensions**.